index.js

```diff
@@ -47,12 +66,18 @@ Piece.prototype.get = function (i) {
 
 Piece.prototype.set = function (i, data, source) {
   if (!this.init()) return false
-  if (!this._buffer[i]) {
-    this._buffered++
-    this._buffer[i] = data
-    this.missing -= data.length
-    if (this.sources.indexOf(source) === -1) {
-      this.sources.push(source)
+  var len = data.length
+  var blocks = Math.ceil(len / BLOCK_LENGTH)
+  for (var j = 0; j < blocks; j++) {
+    if (!this._buffer[i + j]) {
+      var offset = j * BLOCK_LENGTH
+      var splitData = data.slice(offset, offset + BLOCK_LENGTH)
+      this._buffered++
+      this._buffer[i + j] = splitData
+      this.missing -= splitData.length
+      if (this.sources.indexOf(source) === -1) {
+        this.sources.push(source)
+      }
```

Comment:
Nevermind. It needs to stay inside the if-statement. Will leave it as-is for now.